osu.Game/Screens/Select/BeatmapDetails.cs

```diff
@@ -0,0 +1,513 @@
+                description.Text = beatmap.Version;
+                source.Text = beatmap.Metadata.Source;
+                tags.Children = beatmap.Metadata.Tags?.Split(' ').Select(text => new OsuSpriteText
+                {
+                    Text = text,
+                    Font = "Exo2.0-Regular",
+                });
+
+                circleSize.Value = beatmap.Difficulty.CircleSize;
+                drainRate.Value = beatmap.Difficulty.DrainRate;
+                overallDifficulty.Value = beatmap.Difficulty.OverallDifficulty;
+                approachRate.Value = beatmap.Difficulty.ApproachRate;
+                stars.Value = (float)beatmap.StarDifficulty;
+            }
+        }
+
+        private List<int> ratings;
+
+        public IEnumerable<int> Ratings
+        {
+            get
+            {
+                return ratings;
+            }
+            set
+            {
+                ratings = value.ToList();
+                negativeRatings.Text = ratings.GetRange(0, 5).Sum().ToString();
+                positiveRatings.Text = ratings.GetRange(5, 5).Sum().ToString();
+                ratingsBar.Length = (float)ratings.GetRange(0, 5).Sum() / ratings.Sum();
+
+                ratingsGraph.Values = ratings.Select(rating => (float)rating);
+            }
+        }
+
+        private List<int> retries = Enumerable.Repeat(0, 100).ToList();
+
+        public IEnumerable<int> Retries
+        {
+            get
+            {
+                return retries;
+            }
+            set
+            {
+                retries = value.ToList();
+                calcRetryAndFailGraph();
+            }
+        }
+
+        private List<int> fails = Enumerable.Repeat(0, 100).ToList();
+
+        public IEnumerable<int> Fails
+        {
+            get
+            {
+                return fails;
+            }
+            set
+            {
+                fails = value.ToList();
+                calcRetryAndFailGraph();
+            }
+        }
+
+        private void calcRetryAndFailGraph()
+        {
+            failGraph.Values = fails.Select(fail => (float)fail);
+            retryGraph.Values = retries.Select((retry, index) => (float)retry + fails[index]);
+        }
+
+        public BeatmapDetails()
+        {
+            Children = new Drawable[]
+            {
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = Color4.Black,
+                    Alpha = 0.5f,
+                },
+                new FillFlowContainer()
+                {
+                    Anchor = Anchor.TopRight,
+                    Origin = Anchor.TopRight,
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y,
+                    Width = 0.4f,
+                    Direction = FillDirection.Vertical,
+                    Padding = new MarginPadding(10) { Top = 25 },
+                    Children = new Drawable[]

```

Comment:
All of the elements in here should hide if they have no data. e.g. the rating spread should hide if there's no ratings.

All elements following hidden elements should naturally move up once their alpha hits 0, so you'll want to use LayoutDuration/LayoutEasing on this FillFlow.